File: src/servers/rpc_client.ts
import WebSocket from "ws";

export class RpcClient {
  private ws: WebSocket | null = null;
  private pendingRequests = new Map<
    string,
    { resolve: Function; reject: Function; method: string }
  >();
  private messageId = 0;

  /**
   * Generate a unique ID for requests
   */
  private generateId(): string {
    return `${Date.now()}_${this.messageId++}`;
  }

  /**
   * Connect to the Flutter RPC server
   */
  async connect(host: string, port: number, path: string): Promise<void> {
    const readyState = this.ws?.readyState;
    console.log(`readyState: ${readyState}`);

    // Only return early if the WebSocket is in OPEN state
    if (readyState === WebSocket.OPEN) {
      console.log(`Already connected to Flutter RPC server`);
      return Promise.resolve();
    }

    // If WebSocket exists but is not open, close and recreate it
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }

    // Create new WebSocket connection
    return new Promise((resolve, reject) => {
      const wsUrl = `ws://${host}:${port}${path}`;
      this.ws = new WebSocket(wsUrl);
      console.log(`Connecting to Flutter RPC server at ${wsUrl}`);

      this.ws.onopen = () => {
        console.log(`Connected to Flutter RPC server at ${wsUrl}`);
        resolve();
      };

      this.ws.onerror = (error) => {
        console.error(`WebSocket error:`, error);
        reject(error);
      };

      this.ws.onclose = () => {
        console.log(`Disconnected from Flutter RPC server`);
        this.ws = null;
      };

      this.ws.onmessage = (event) => {
        try {
          const response = JSON.parse(event.data.toString());

          if (response.id) {
            const request = this.pendingRequests.get(response.id);
            if (request) {
              if (response.error) {
                request.reject(new Error(response.error.message));
              } else {
                request.resolve(response.result);
              }
              this.pendingRequests.delete(response.id);
            }
          }
        } catch (error) {
          console.error("Error parsing WebSocket message:", error);
        }
      };
    });
  }

  /**
   * Call a method on the Flutter RPC server
   */
  async callMethod(
    method: string,
    params: Record<string, unknown> = {}
  ): Promise<unknown> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error(
        `Not connected to Flutter RPC server ${this.ws?.readyState}`
      );
    }

    const id = this.generateId();

    const request = {
      jsonrpc: "2.0",
      id,
      method,
      params,
    };

    return new Promise((resolve, reject) => {
      this.pendingRequests.set(id, { resolve, reject, method });
      this.ws!.send(JSON.stringify(request));
    });
  }

  /**
   * Disconnect from the Flutter RPC server
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}

File: devtools_mcp_extension/lib/services/rpc_client.dart
// ignore_for_file: avoid_catches_without_on_clauses

import 'dart:convert';

import 'package:flutter_mcp_extension/common_imports.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

class RpcClient extends ChangeNotifier {
  WebSocketChannel? _channel;
  final Map<String, Function> _methods = {};
  var _connected = false;

  bool get connected => _connected;

  /// Returns a list of registered method names
  List<String> get registeredMethods => _methods.keys.toList();

  // Register RPC methods that can be called by the TypeScript server
  void registerMethod(final String methodName, final Function handler) {
    _methods[methodName] = handler;
    notifyListeners();
  }

  Future<void> connect({
    required final int port,
    required final String host,
    required final String path,
  }) async {
    await _startClient(port, host, path);
  }

  Future<void> _startClient(
    final int port,
    final String host,
    final String path,
  ) async {
    try {
      final wsUrl = 'ws://$host:$port/$path';
      print('Client connecting to $wsUrl');

      // Use the WebSocketChannel to connect to the server
      final channel = WebSocketChannel.connect(Uri.parse(wsUrl));

      // Listen to the incoming messages
      channel.stream.listen(
        (final message) async {
          await _handleRpcCall(channel, message);
        },
        onDone: () {
          _channel = null;
          _connected = false;
          notifyListeners();
          print('Disconnected from WebSocket server');

          // Try to reconnect after a delay
          Future.delayed(const Duration(seconds: 2), () {
            if (!_connected) {
              print('Attempting to reconnect...');
              connect(port: port, host: host, path: path);
            }
          });
        },
        onError: (final error) {
          _channel = null;
          _connected = false;
          notifyListeners();
          print('Error with WebSocket connection: $error');

          // Try to reconnect after a delay
          Future.delayed(const Duration(seconds: 2), () {
            if (!_connected) {
              print('Attempting to reconnect after error...');
              connect(port: port, host: host, path: path);
            }
          });
        },
      );

      _channel = channel;
      _connected = true;
      notifyListeners();
      print('Connected to WebSocket server');
    } catch (e) {
      print('Failed to connect to WebSocket server: $e');
      // Try to reconnect after a delay
      Future.delayed(const Duration(seconds: 2), () {
        if (!_connected) {
          print('Attempting to reconnect after failure...');
          connect(port: port, host: host, path: path);
        }
      });
    }
  }

  Future<void> _handleRpcCall(
    final WebSocketChannel channel,
    final rawMessage,
  ) async {
    try {
      final Map<String, dynamic> message = json.decode(rawMessage.toString());
      final String id = message['id'];
      final String method = message['method'];
      final Map<String, dynamic> params = message['params'] ?? {};

      print('Received RPC call: $method (ID: $id)');

      if (_methods.containsKey(method)) {
        // Execute the method and get result
        await _executeRpcMethod(channel, id, method, params);
      } else {
        // Method not found
        _sendResponse(channel, id, null, {
          'code': -32601,
          'message': 'Method not found',
        });
      }
    } catch (e, stackTrace) {
      print('Error processing message: $e $stackTrace');
      // Parse error
      _sendResponse(channel, '0', null, {
        'code': -32700,
        'message': 'Parse error',
      });
    }
  }

  Future<void> _executeRpcMethod(
    final WebSocketChannel channel,
    final String id,
    final String method,
    final Map<String, dynamic> params,
  ) async {
    try {
      final handler = _methods[method]!;

      // Execute handler and get result
      final result = await Function.apply(handler, [params]);

      // Send response back to client
      _sendResponse(channel, id, result, null);
    } catch (e, stackTrace) {
      print('Error executing method $method: $e $stackTrace');
      _sendResponse(channel, id, null, {
        'code': -32603,
        'message': 'Internal error: $e',
      });
    }
  }

  void _sendResponse(
    final WebSocketChannel channel,
    final String id,
    final result,
    final error,
  ) {
    final response = {
      'jsonrpc': '2.0',
      'id': id,
      if (result != null) 'result': result,
      if (error != null) 'error': error,
    };

    // WebSocketChannel has a unified API for both web and non-web
    channel.sink.add(json.encode(response));
  }

  Future<void> disconnect() async {
    await _channel?.sink.close();
  }
}
